credmark/cmf/types/ledger_contract.py

```diff
@@ -0,0 +1,283 @@
+from enum import Enum
+from typing import List, Tuple, Union
+
+import credmark.cmf.model
+from credmark.cmf.model.errors import ModelInputError, ModelRunError
+
+from .abi import ABI
+from .ledger import ColumnField, LedgerModelOutput, LedgerTable
+from .ledger_query import LedgerQueryBase
+
+
+class ContractFunctionsTable(LedgerTable):
+    """
+    Contract functions ledger data table
+    Column names common to all functions
+    Function-specific columns are added from ABI.
+    """
+
+    CONTRACT_ADDRESS = ColumnField('contract_address')
+    """"""
+    TXN_BLOCK_NUMBER = ColumnField('txn_block_number')
+    """"""
+    TXN_HASH = ColumnField('txn_hash')
+    """"""
+    TXN_INDEX = ColumnField('txn_index')
+    """"""
+    SUCCESS = ColumnField('success')  # boolean indicating txn was successful or not
+    """"""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+
+class ContractEventsTable(LedgerTable):
+    """
+    Contract events ledger data table
+    Column names common to all events
+    Event-specific columns are added from ABI.
+    """
+
+    CONTRACT_ADDRESS = ColumnField('contract_address')
+    """"""
+    EVT_BLOCK_NUMBER = ColumnField('evt_block_number')
+    """"""
+    EVT_HASH = ColumnField('evt_tx_hash')
+    """"""
+    EVT_INDEX = ColumnField('evt_index')
+    """"""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+
+class ContractEntityType(Enum):
+
+    """"""
+    FUNCTIONS = 'functions'
+    """"""
+    EVENTS = 'events'
+    """"""
+
+
+class ContractEntityQuery(LedgerQueryBase):
+    """
+    Used by :class:`~credmark.cmf.types.ledger.ContractLedger` to query a
+    contract's function or event data.
+    You do not need to create an instance yourself.
+
+    Access an instance with ``contract.ledger.functions.contractFunctionName()``
+    or ``contract.ledger.events.contractEventName()`` where ``contractFunctionName``
+    and ``contractEventName`` are the actual names of functions and events
+    of the contract.
+
+    See the :meth:`~credmark.cmf.types.ledger.ContractLedger.ContractEntity.__call__`
```

Comment:
This needs to be corrected for the `select` call... `~credmark.cmf.types.ledger_contract. ContractEntityQuery.select`